[Hub Generated] Review request for Microsoft.ResourceGraph to add version preview/2020-04-01-preview

Yobretaw opened the pull request:
This PR
- Separated out resource change APIs from 2020-04-01-preview resourcegraph.json into a dedicated swagger file.
- Added a new API `/partner/microsoft.resourcegraph/resourcesHistory`
  - This is for the S360 item https://portal.azure-devex-tools.com/amekpis/completeness/detail?errorId=21A3A9D7-34CF-4625-B493-6A600E0AD60C
- **[BreakingChange]** Rename `eTag` to `etag` as mentioned in #9466 
- **[BreakingChange]** Added missing `resourceId` field to `ResourceChangeData` model, and added missing `snapshotId` field to `ResourceSnapshotData` model
  - These are for the S360 item https://portal.azure-devex-tools.com/amekpis/correctness/detail?errorId=6208C7B6-5F04-46D1-BE0B-42DDA74BA1B6

This is a PR generated at [OpenAPI Hub](https://aka.ms/openapihub). You can view your [work branch via this link](https://portal.azure-devex-tools.com/branch/Yobretaw/azure-rest-api-specs/dev-resourcegraph-Microsoft.ResourceGraph-2020-04-01-preview...master/).

### Contribution checklist:
- [x] I commit to follow the [Breaking Change Policy](http://aka.ms/bcforapi) of “no breaking changes
- [x] I have reviewed the [documentation](https://aka.ms/ameonboard) for the workflow.
- [ ] [Validation tools](https://aka.ms/swaggertools) were run on swagger spec(s) and errors have all been fixed in this PR. [How to fix?](https://aka.ms/ci-fix)

If any further question about AME onboarding or validation tools, please view the [FAQ](https://aka.ms/faqinprreview).

### ARM API Review Checklist
- [x] Ensure to check this box if one of the following scenarios meet updates in the PR, so that label “**WaitForARMFeedback**” will be added automatically to involve ARM API Review. Failure to comply may result in delays for manifest application. Note this does not apply to data plane APIs, all “removals” and “adding a new property” no more require ARM API review.
  - Adding new API(s)
  - Adding a new API version
  - Adding a new service

- [ ] If you are blocked on ARM review and want to get the PR merged with urgency, please get the ARM oncall for reviews (*RP Manifest Approvers* team under <ins>Azure Resource Manager service</ins>) from IcM and reach out to them. 

### Breaking Change Review Checklist 
If there are following updates in the PR, ensure to request an approval from API Review Board as defined in the [Breaking Change Policy](http://aka.ms/bcforapi). 

- [ ] Removing API(s) in stable version
- [ ] Removing properties in stable version
- [ ] Removing API version(s) in stable version
- [ ] Updating API in stable version with Breaking Change Validation errors
- [x] Updating API(s) in preview over 1 year

Please follow the link to find more details on [PR review process](https://aka.ms/SwaggerPRReview).

Yobretaw committed:
Add missing subscriptionId and resourceGroup

Yobretaw committed:
Update 2020-04-01-preview

Commit:
Split 2020-04-01-preview

Commit:
Merge branch 'patch-1' into dev-resourcegraph-Microsoft.ResourceGraph-2020-04-01-preview

AutorestCI commented:
/azp run automation - sdk

Yobretaw commented:
@allenjzhang There are no new API added in this PR. We are just separating them into separate files.

allenjzhang commented on specification/resourcegraph/resource-manager/Microsoft.ResourceGraph/preview/2018-09-01-preview/graphquery.json:
```diff
@@ -267,6 +267,16 @@
           "readOnly": true,
           "description": "Azure resource name. This is GUID value. The display name should be assigned within properties field."
         },
+        "subscriptionId": {
```
I do not understand the change here. This "Resource" definition was more or less aligned with  standard "TrackedResource" definition. But now you are adding subscription ID and resource group here, it is messing up this base definition. I would suggest this to go on to the specific query definition.

Also in terms of the name `GraphQueryResource`, if feels more like `XXXQuery` rather a `XXXResource`. Anyway, I guess that boat has sailed.

Yobretaw replied:
Actually you are correct.. We are returning extra fields in the RP. Let me revert this change and fix our RP side.

Yobretaw replied:
Reverted changes for subscriptionId/resourceGroup. Also fixed a casing issue for `etag` field.

Commit:
Reverting changes for subscriptionId/resourceGroup; Fix casing for etag field.

AutorestCI commented:
/azp run automation - sdk

allenjzhang commented on specification/resourcegraph/resource-manager/Microsoft.ResourceGraph/preview/2020-04-01-preview/resourcechanges.json:
```diff
@@ -0,0 +1,405 @@
+    "/providers/Microsoft.ResourceGraph/resourceChanges": {
+      "post": {
+        "description": "List changes to a resource for a given time interval.",
+        "operationId": "ResourceChanges",
```
OperationId should be in Noun_Verb format. Suggest `ResourceChanges_List`.  If you have/plan additional List at different scope either now or later, you could add a `ByXXX` suffix here.  Such as `ResourceChanges_ListByTenant` as this is across all subs.

Yobretaw replied:
Thanks for the suggestions. For now we'd like to keep it as-is, but we'll definitely consider revisiting the operationId in future releases.

allenjzhang commented on specification/resourcegraph/resource-manager/Microsoft.ResourceGraph/preview/2020-04-01-preview/resourcechanges.json:
```diff
@@ -0,0 +1,405 @@
+  },
+  "paths": {
+    "/providers/Microsoft.ResourceGraph/resourceChanges": {
+      "post": {
```
Feels this can be supported with GET particularly you have $top, $skip which is typical of query string parameters. But I will leave ARM reviewer to chime in here.

majastrz replied:
Unfortunately no because GET doesn't allow a request body. (There's a "body" parameter below.)

---
In reply to: [508220282](https://github.com/Azure/azure-rest-api-specs/pull/11070#discussion_r508220282) [](ancestors = 508220282)

majastrz replied:
Well technically speaking some of the parameters in the request body could probably be moved to query string, but I'm not going to insist on that given that the rest of the RP is modeled using POST.

---
In reply to: [552319239](https://github.com/Azure/azure-rest-api-specs/pull/11070#discussion_r552319239) [](ancestors = 552319239,508220282)

allenjzhang reviewed (commented):
Can you indicate whether there are changes to resourcechanges.json OR a simple restructure of the swagger with no changes? 

I understand this is not new. So please treat the following comment as guidelines as we don't want to introduce breaking changes. But nevertheless, these guidelines should be aware of and strive to fix. Let's evaluate the impact on SDK breaking changes to see if we can make these changes. @akning-ms and @markcowl for FYI.

Yobretaw commented:
> Can you indicate whether there are changes to resourcechanges.json OR a simple restructure of the swagger with no changes?

This PR simply restructures the swagger without adding/removing/modifying any APIs or models.

Commit:
Adding missing ResourcesHistory API to swagger

Commit:
Fix API definition; Add an example

Commit:
Fix multiple issues

1) Rename PropertyChangeType to ChangeType; 2) Add resourceId ResourceChangeData; 3) Add snapshotId to ResourceSnapshotData; 4) Fix model name in resourceshistory.json

Commit:
Fix linter/model issues

Commit:
Fix etag casing more places

fenwickt commented on specification/resourcegraph/resource-manager/Microsoft.ResourceGraph/preview/2020-04-01-preview/resourcechanges.json:
```diff
@@ -0,0 +1,413 @@
+              "readOnly": true,
+              "type": "array",
+              "items": {
+                "$ref": "#/definitions/ErrorResponse"
```
by any chance should this be #/definitions/ErrorDetails?

Yobretaw replied:
Good catch! Somehow when I messed it up when moving the API to a new file..

Commit:
Fix ErrorResponse modeling in resourcechanges.json and resourcehistory.json

Commit:
Add default response for Operations API

Commit:
Revert renaming of PropertyChangeType

Commit:
Add changeType to swagger

Commit:
Revert "Add changeType to swagger"

This reverts commit 5bfd7fbd9492abc96076a459f7ca914e44bffc9a.

majastrz reviewed (approved):
Signed off from ARM side.

Yobretaw commented:
@allenjzhang Both BreakingChange and ARM API Reviewer have signed off. I took a look into LintDiff but errors there are expected. Could you please take a look?

Commit:
Fix swagger for resourcesHistory
